### DISCUSSION TITLE
Why does vcpkg_cmake_config_fixup move a package's config file to share?

### DISCUSSION THREAD

**Discussion by nickelpro** [microsoft/vcpkg#33108] [Q&A] [answered]
There are a number of valid places a package can choose to install its config file, and some packages make the reasonable assumption that their config file will stay where they left it.

`vcpkg_cmake_config_fixup` will move the config file to `share/` (if supplied with the correct CONFIG_PATH), which breaks some `Config.cmake`s which must then be fixed with patches or `vcpkg_replace_string`s.

Notably, the entire LLVM ecosystem, both in-tree and out-of-tree tooling, assumes their cmake configs are installed to `lib/cmake` and ascend two levels from `${CMAKE_CURRENT_LIST_DIR}` to find the prefix.

Is this a reasonable thing to do on LLVM's part? Maybe, maybe not. But they're certainly justified in thinking it will work, as CMake doesn't provide a native way to retarget the relative location of the config file within the install tree.

So before I go advocating to make other's CMLs more compatible with vcpkg, why does vcpkg do this? And is there a better way that doesn't involve screwing up other project's valid assumptions about the layout of the install tree?

**Comment by Neumann-A** [comment]
> So before I go advocating to make other's CMLs more compatible with vcpkg, why does vcpkg do this? And is there a better way that doesn't involve screwing up other project's valid assumptions about the layout of the install tree?

Because multi config. CMake assumes different configs to be installable into the same prefix which is simply not the case if the library does not specify a different suffix between release/debug. As such to enable multi config in cmake vcpkg has to install into different prefixes `<installed>/` and `<installed>/debug` and then move the cmake files into a single directory accordingly. Do not force `share/` upstream unless the package is header only. 
`lib-<arch>/cmake`  is the correct place for everything abi relevant however for vcpkg it does not matter since abi is ensured by the triplet name within the path already.

**Reply by traversaro** [reply]
To complement @Neumann-A's answer, this is an old issue in which this was discussed: https://github.com/microsoft/vcpkg/issues/77 .

**Reply by nickelpro** [reply]
I don't think this answers the question though. I totally understand why we need to move the cmake files into a single directory (specifically the `*-release.cmake` and `*-debug.cmake` files properly to populate targets).

My question is why is that directory `share/`? Why can't vcpkg populate `lib/cmake` with the config files? Or whatever path is passed to `CONFIG_PATH`?

They just need to be in be in the same directory. If we could choose which directory that was (or just use whatever directory the current project has designated), there would be no need to patch any upstream packages

**Reply by Neumann-A** [reply]
> They just need to be in be in the same directory. If we could choose which directory that was (or just use whatever directory the current project has designated), there would be no need to patch any upstream packages

You don't need to patch you use `vcpkg_cmake_config_fixup` to fix the location

**Reply by nickelpro** [reply]
Again, `vcpkg_cmake_config_fixup` relocating, ie, MLIRConfig.cmake (or any LLVM config file) to `share/` breaks those files and requires either a patch or a string replacement because they make assumptions about their position relative to the rest of the install tree.

Specifically this is standard practice in LLVM projects:

```
get_filename_component(LLVM_INSTALL_PREFIX "${CMAKE_CURRENT_LIST_FILE}" PATH)
get_filename_component(LLVM_INSTALL_PREFIX "${LLVM_INSTALL_PREFIX}" PATH)
get_filename_component(LLVM_INSTALL_PREFIX "${LLVM_INSTALL_PREFIX}" PATH)
get_filename_component(LLVM_INSTALL_PREFIX "${LLVM_INSTALL_PREFIX}" PATH)

# ...
set(LLVM_CMAKE_DIR "${LLVM_INSTALL_PREFIX}/lib/cmake/llvm")
```

Both of these "reasonable" idioms break with the move to `share/`

**Reply by Neumann-A** [reply]
Again that is just the price vcpkg has to pay for the layout. You are however free to install into `share/cmake/MLIR/` to at least fix one of those.

`set(LLVM_CMAKE_DIR "${LLVM_INSTALL_PREFIX}/lib/cmake/llvm")` is illegal. (Since it is MLIR not LLVM even though MLIR is part of LLVM, the correct approach here is to use find_dependency(LLVM) which pulls in that variable. )

**Reply by nickelpro** [reply]
The example code is from LLVM proper, not MLIR. I shouldn't have used disconnected examples, apologies for the confusion.

> Again that is just the price vcpkg has to pay for the layout.

And again, my question is why is vcpkg paying this price?

What would break if:

```
vcpkg_cmake_config_fixup(
  PACKAGE_NAME llvm
  CONFIG_PATH lib/cmake/llvm
)
```

Resulted in the following install tree:

```
${_IMPORT_PREFIX}/lib/cmake/llvm/LLVMConfig.cmake
${_IMPORT_PREFIX}/lib/cmake/llvm/LLVMExports.cmake
${_IMPORT_PREFIX}/lib/cmake/llvm/LLVMExports-debug.cmake
${_IMPORT_PREFIX}/lib/cmake/llvm/LLVMExports-release.cmake
```

LLVM wants to live in `lib/cmake`, vcpkg wants all the `*-${CMAKE_BUILD_TYPE}.cmake`s to live in the same directory.  These things aren't incompatible, am I missing something?

**Reply by Neumann-A** [reply, marked as answer]
a) consistency

> LLVM wants to live in lib/cmake

b) as far as i remember there is a LLVM cmake variable which actually controls where the cmake configs go (`LLVM_CONFIG_CMAKE_DIR`/ `LLVM_INSTALL_PACKAGE_DIR`). So vcpkg can just set that correctly and move the files.

**Reply by nickelpro** [reply]
Consistency is a pretty terrible reason to break valid CMLs, but it's an answer I suppose

**Reply by Neumann-A** [reply]
And c) the variable is still wrong since it makes no sense to calculate the path if it is just cmake_current_list_dir. 
In general I have yet to see good cmake configs where any of these variables are really needed and not just dead or superflous code.

**Reply by nickelpro** [reply]
I 100% agree with you on that:
> Is this a reasonable thing to do on LLVM's part? Maybe, maybe not. But they're certainly justified in thinking it will work

I think it's completely pointless, but they rely on it, CMake supports it, and so vcpkg shouldn't break it just to get all the config files nicely organized.

That said, not my monkeys, not my circus. The question was along the lines of "is this a vcpkg bug or is this intended behavior?" and that's answered.
